### PR TITLE
Book a ride, get pending rides and accept a ride APIs

### PR DESCRIPTION
**Pull request by nikhilleo10**
### Ticket Link
Book a ride API
List all pending rides API
Accept a ride API
---

### Related Links

---

### Description
- Customer can request a ride by providing pickup and drop location
- Driver can get list of all the pending rides that he/she can accept
- Driver can accept a ride by providing his/her id and the ride id
---

### Steps to Reproduce / Test

---

---

### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added]
- [ ] Relevant documentation is changed or added (and PR referenced)

### GIF's

---

### CODE REVIEW

File: lib/daos/ridesDao.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+export const bookARide = async (pickUpLocation, dropLocation, custId) => {
+  try {
+    const ridesData = {
+      pickupLoc:pickUpLocation,
+      dropLoc: dropLocation,
+      dateOfRide: moment(moment.now()).format('YYYY-MM-DD'),
+      bookingTime: moment(moment.now()).format("YYYY-MM-DD HH:mm:ss"),
+      custId,
+      created_at: moment(moment.now()).format("YYYY-MM-DD HH:mm:ss")
```
Comment (aseer-ws) [inline lib/daos/ridesDao.js:14]: Why we are storing duplicate data in different columns `dateOfRide`, `bookingTime`, `created_at`

File: lib/routes/rides/routes.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
+import get from 'lodash/get';
+import { notFound, badImplementation, badRequest } from '@utils/responseInterceptors';
+import Joi from 'joi';
+import { bookARide, getPendingRides, assignDriverToRide } from '@daos/ridesDao';
+import { customerIdSchema } from '@utils/validationUtils';
+
+export default [
+  {
+    method: 'POST',
+    path: '/book/{customerId}',
```
Comment (aseer-ws) [inline lib/routes/rides/routes.js]:
```suggestion
    path: '/',
```
Lets send the `customerId` in body itself

File: lib/routes/rides/routes.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+          }
+          return notFound('No Pending Rides found');
+      } catch (error) {
+        throw badImplementation(error.message)
+      }
+    },
+  },
+  {
+    method: 'POST',
+    path: '/accept/{driverId}/{rideId}',
```
Comment (aseer-ws) [inline lib/routes/rides/routes.js]:
```suggestion
    path: '/{rideId}/accept',
```
Let's send `driverId` in body.

File: lib/routes/rides/routes.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    options: {
+      description: 'Accept a ride',
+      notes: 'Accept a ride API',
+      tags: ['api', 'rides'],
+      validate: {
+        params: Joi.object({
+          driverId: Joi.number().required().error(new Error('Invalid Driver Id.')),
```
Comment (aseer-ws) [inline lib/routes/rides/routes.js]:
```suggestion
          driverId: numberSchema.error(new Error('Invalid Driver Id.')),
```

File: lib/daos/ridesDao.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    return await models.requested_rides.create(ridesData);
```
Comment (aseer-ws) [inline lib/daos/ridesDao.js]:
```suggestion
    return await models.requestedRides.create(ridesData);
```
Let's keep model name camelcase

File: lib/routes/rides/routes.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      description: 'Book a ride for a user',
+      notes: 'Book a ride for a user API',
+      tags: ['api', 'rides'],
+      validate: {
+        params: Joi.object({
+          customerId: customerIdSchema 
+        }),
+        payload: Joi.object({
+          pickUpLocation: Joi.string().trim().required().error(new Error('Pickup location is required.')),
+          dropLocation: Joi.string().trim().required().error(new Error('Drop location is required.'))
```
Comment (aseer-ws) [inline lib/routes/rides/routes.js]: Let's move repetitive schema into `validationUtils.js`
```js
const trimmedStringSchema = Joi.string().trim().required()
```
And while implementing
```js
trimmedStringSchema.error(new Error('Drop location is required.'))
```

**Review by aseer-ws — CHANGES_REQUESTED** [review]

File: lib/daos/ridesDao.js (hunk 1 above)
Comment (nikhilleo10) [reply lib/daos/ridesDao.js:14]: - Date of ride is only date on which the ride is scheduled.
- Booking time we are taking the current time at which the API is called.
- Created at will also be the same.

File: lib/routes/rides/routes.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            })
+          }
+          return notFound('No Pending Rides found');
+      } catch (error) {
+        throw badImplementation(error.message)
+      }
+    },
+  },
+  {
+    method: 'POST',
```
Comment (aseer-ws) [inline lib/routes/rides/routes.js:79]:
```suggestion
    method: 'PATCH',
```

**Review by aseer-ws — APPROVED** [review]

**Pull request merged**